atc/atccmd/command.go

```diff
@@ -1518,14 +1577,114 @@ func (cmd *RunCommand) constructHTTPHandler(
+				Handler:    webMux,
+				Middleware: middleware,
 			},
 		},
 	}
 
 	return httpHandler
 }
 
+func (cmd *RunCommand) constructLegacyHandler(
+	logger lager.Logger,
+) (http.Handler, error) {
+	return legacyserver.NewLegacyServer(&legacyserver.LegacyConfig{
+		Logger: logger.Session("legacy"),
+	})
+}
+
+func (cmd *RunCommand) constructAuthHandler(
+	logger lager.Logger,
+	storage storage.Storage,
+) (http.Handler, error) {
+
+	issuerPath, _ := url.Parse("/sky/issuer")
+	redirectPath, _ := url.Parse("/sky/callback")
+
+	issuerURL := cmd.ExternalURL.URL.ResolveReference(issuerPath)
+	redirectURL := cmd.ExternalURL.URL.ResolveReference(redirectPath)
+
+	signingKey, err := loadOrGenerateSigningKey(cmd.Auth.AuthFlags.SigningKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add public fly client
+	cmd.Auth.AuthFlags.Clients["fly"] = "Zmx5"
+
+	dexServer, err := dexserver.NewDexServer(&dexserver.DexConfig{
+		Logger:      logger.Session("dex"),
+		Users:       cmd.Auth.AuthFlags.LocalUsers,
+		Clients:     cmd.Auth.AuthFlags.Clients,
+		Expiration:  cmd.Auth.AuthFlags.Expiration,
+		IssuerURL:   issuerURL.String(),
+		RedirectURL: redirectURL.String(),
+		WebHostURL:  "/sky/issuer",
+		SigningKey:  signingKey,
+		Storage:     storage,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return dexServer, nil
+}
+
+func (cmd *RunCommand) constructLoginHandler(
+	logger lager.Logger,
+	httpClient *http.Client,
+	middleware token.Middleware,
+) (http.Handler, error) {
+
+	authPath, _ := url.Parse("/sky/issuer/auth")
+	tokenPath, _ := url.Parse("/sky/issuer/token")
+	redirectPath, _ := url.Parse("/sky/callback")
+
+	authURL := cmd.ExternalURL.URL.ResolveReference(authPath)
+	tokenURL := cmd.ExternalURL.URL.ResolveReference(tokenPath)
+	redirectURL := cmd.ExternalURL.URL.ResolveReference(redirectPath)
+
+	endpoint := oauth2.Endpoint{
+		AuthURL:   authURL.String(),
+		TokenURL:  tokenURL.String(),
+		AuthStyle: oauth2.AuthStyleInHeader,
+	}
+
+	oauth2Config := &oauth2.Config{
+		Endpoint:     endpoint,
+		ClientID:     cmd.Server.ClientID,
+		ClientSecret: cmd.Server.ClientSecret,
+		RedirectURL:  redirectURL.String(),
+		Scopes:       []string{"openid", "profile", "email", "federated:id", "groups"},
+	}
+
+	skyServer, err := skyserver.NewSkyServer(&skyserver.SkyConfig{
+		Logger:          logger.Session("sky"),
+		TokenMiddleware: middleware,
+		OAuthConfig:     oauth2Config,
+		HTTPClient:      httpClient,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return skyserver.NewSkyHandler(skyServer), nil
+}
+
+func (cmd *RunCommand) constructTokenVerifier(httpClient *http.Client) accessor.Verifier {
+
+	publicKeyPath, _ := url.Parse("/sky/issuer/keys")
+	publicKeyURL := cmd.ExternalURL.URL.ResolveReference(publicKeyPath)
+
+	validClients := []string{"fly"}
+	for clientId, _ := range cmd.Auth.AuthFlags.Clients {
+		validClients = append(validClients, clientId)
+	}
+
+	return accessor.NewVerifier(httpClient, publicKeyURL, validClients)
+}
```

Comment:
Could get the verifier to fetch the the `KeySet` here ?

Reply:
This is part of the startup path. I'd rather fetch the keyset lazily on first request than block on startup.